File: backend/app/games/minion_clash/js/battle/MovementSystem.js
import { GameConfig } from '../config/GameConfig.js';
import { EntityKind } from '../entities/Entity.js';

/**
 * Soft-body separation: pushes overlapping units apart so blobs don't fully
 * stack. Skips towers (immovable) and projectiles. Cheap O(n*k) using spatial
 * index queries with small radius.
 */
export class MovementSystem {
    constructor(spatial) { this._spatial = spatial; }

    update(entityManager, dt) {
        const force = GameConfig.BATTLE.SEPARATION_FORCE;
        for (const e of entityManager.list()) {
            if (e.kind !== EntityKind.UNIT && e.kind !== EntityKind.HERO) continue;
            this._separateOne(e, force, dt);
            this._clampX(e);
        }
    }

    _separateOne(e, force, dt) {
        const queryR = e.radius * 2 + 4;
        const others = this._spatial.queryAll(e.x, e.y, queryR);
        for (const o of others) {
            if (o === e) continue;
            if (o.kind === EntityKind.PROJECTILE || o.kind === EntityKind.TOWER) continue;
            this._pushApart(e, o, force, dt);
        }
    }

    _pushApart(e, o, force, dt) {
        const dx = e.x - o.x, dy = e.y - o.y;
        const distSq = dx * dx + dy * dy;
        const min = e.radius + o.radius;
        if (distSq >= min * min || distSq < 0.0001) return;
        const dist = Math.sqrt(distSq);
        const overlap = (min - dist) / min;
        const push = force * overlap * dt;
        e.x += (dx / dist) * push;
        e.y += (dy / dist) * push;
    }

    _clampX(e) {
        if (e.x < 12) e.x = 12;
        else if (e.x > GameConfig.VIEW_WIDTH - 12) e.x = GameConfig.VIEW_WIDTH - 12;
    }
}

File: backend/app/games/minion_clash_be/simulation/movement_system.py
"""
MovementSystem — soft-body separation. Mirrors `MovementSystem.js`.

Pushes overlapping units/heroes apart so blobs don't fully stack.
Skips towers (immovable) and projectiles. O(n*k) via spatial queries.
"""

from __future__ import annotations

from . import arena
from .entities import KIND_HERO, KIND_PROJECTILE, KIND_TOWER, KIND_UNIT
from .entity_manager import EntityManager
from .spatial_index import SpatialIndex


class MovementSystem:
    def __init__(self, spatial: SpatialIndex):
        self._spatial = spatial

    def update(self, em: EntityManager, dt: float) -> None:
        force = arena.SEPARATION_FORCE
        for e in em.list():
            if e.kind != KIND_UNIT and e.kind != KIND_HERO:
                continue
            self._separate_one(e, force, dt)
            self._clamp_x(e)

    def _separate_one(self, e, force: float, dt: float) -> None:
        query_r = e.radius * 2 + 4
        for o in self._spatial.query_all(e.x, e.y, query_r):
            if o is e:
                continue
            if o.kind == KIND_PROJECTILE or o.kind == KIND_TOWER:
                continue
            self._push_apart(e, o, force, dt)

    @staticmethod
    def _push_apart(e, o, force: float, dt: float) -> None:
        dx = e.x - o.x
        dy = e.y - o.y
        dist_sq = dx * dx + dy * dy
        min_d = e.radius + o.radius
        if dist_sq >= min_d * min_d or dist_sq < 0.0001:
            return
        dist = dist_sq ** 0.5
        overlap = (min_d - dist) / min_d
        push = force * overlap * dt
        e.x += (dx / dist) * push
        e.y += (dy / dist) * push

    @staticmethod
    def _clamp_x(e) -> None:
        if e.x < 12:
            e.x = 12.0
        elif e.x > arena.VIEW_WIDTH - 12:
            e.x = float(arena.VIEW_WIDTH - 12)
